test/test_modules.py

```diff
@@ -205,6 +206,115 @@ def test_check_inplace(self, device, dtype, module_info):
             output_ip.backward(grad)
             self.assertEqual(input_args[0].grad, input_arg_copy[0].grad)
 
+    def _traverse_obj(self, obj, func):
+        if isinstance(obj, (tuple, list)) or isgenerator(obj):
+            return tuple(self._traverse_obj(o, func) for o in obj)
+        elif isinstance(obj, dict):
+            return {name: self._traverse_obj(o, func) for name, o in obj.items()}
+        elif isinstance(obj, (torch.Tensor, torch.nn.Parameter)):
+            return func(obj)
+
+    def _retain_grad(self, obj):
+        # gradients needs to be retained to check for grad. This is useful when
+        # non-leafs are present in the graph.
+        def inner_retain_grad(obj):
+            if obj.requires_grad:
+                obj.retain_grad()
+        self._traverse_obj(obj, inner_retain_grad)
+
+    def _get_grads(self, obj):
+        def inner_get_grad(obj):
+            if obj.requires_grad:
+                return obj.grad
+        return self._traverse_obj(obj, inner_get_grad)
+
+    def _zero_grad(self, obj):
+        def inner_zero_grad(obj):
+            if obj.grad is not None:
+                obj.grad.detach_()
+                obj.grad.zero_()
+        self._traverse_obj(obj, inner_zero_grad)
```

Comment:
Cool, I like the factored out version :) Should these be moved to utils?

Reply:
I would keep them here and refactor them into `utils` if there are other places in the codebase that have the same pattern?